### PR TITLE
ENH: ndimage.vectorized_filter: make CuPy-compatible

### PR DESCRIPTION
**Pull request by mdhaber**
#### Reference issue
https://github.com/scipy/scipy/pull/22575#discussion_r1976665455

#### What does this implement/fix?
This makes `ndimage.vectorized_filter` CuPy-compatible trivially - the function "just works" when `np` is converted to `xp`, where `xp` is returned by `array_namespace`. Performance compared to CuPy native filters is not terrific, but it still beats NumPy for large jobs, and it might be a nice alternative to `cupyx.scipy.ndimage.generic_filter`, which doesn't accept regular CuPy callables.

#### Additional information
@ev-br rather than creating a CuPy-specific skip-list, I made `vectorized_filter_signature = None`. But I see that we probably do want a CuPy-specific list so arrays from other CPU backends can be converted to a NumPy arrray, filtered, and then converted back. Do I need to put that logic manually in the function? (It didn't seem to be happening automatically when I had the original `vectorized_filter_signatuure`.)

### CODE REVIEW

**Review by ev-br — COMMENTED** [review]
I'd much rather keep the dispatch/delegation logic as centralized as reasonably possible. And minimize the difference between `signal`, `ndimage` and wherever else this sort of `_support_alternative_backends` will be needed in the future.

So currently we only support two kinds of delegation:
- just dispatch to the same function in the `array_module`, as returned by `*_sugnature`
- call the function as is and expect it to convert inputs to numpy. And then postprocess the inputs in the decorator.

If I understand what you're saying, you're looking for a third option:
- if inputs are `cupy`-compatible, call the function and do not wrap either inputs or outputs.

This would actually be great, especially if in the future we'll be able to zero-copy convert GPU inputs to CuPy via dlpack or something. 

So maybe it's about time we generalize the `_signature` framework. I'd still keep as much as possible out of the functions themselves and keep all the delegation logic in `_supports_alternative_backends`.

File: scipy/ndimage/_filters.py [hunk 1]
```diff
@@ -70,13 +73,13 @@ def _vectorized_filter_iv(input, function, size, footprint, output, mode, cval,
     footprinted_function = function
     if size is not None:
         # If provided, size must be an integer or tuple of integers.
-        size = (size,)*input.ndim if np.isscalar(size) else tuple(size)
-        valid = [np.issubdtype(np.asarray(i).dtype, np.integer) and i > 0 for i in size]
+        size = (size,)*input.ndim if xp.isscalar(size) else tuple(size)
+        valid = [xp.issubdtype(xp.asarray(i).dtype, xp.integer) and i > 0 for i in size]
```
Comment (ev-br) [inline scipy/ndimage/_filters.py:77]: Oh, and `xp.issubdtype(..., xp.integer` should probably be `xp.isdtype(..., "integral")` instead.

File: scipy/ndimage/tests/test_filters.py [hunk 2]
```diff
@@ -2766,52 +2766,61 @@ def test_gh_22333():
     assert_array_equal(actual, expected)
 
 
+vf_skip_reason = "`vectorized_filter` requires `sliding_window_view`"
+@pytest.mark.skip_xp_backends('array_api_strict', reason=vf_skip_reason)
+@pytest.mark.skip_xp_backends('dask.array', reason=vf_skip_reason)
+@pytest.mark.skip_xp_backends('torch', reason=vf_skip_reason)
+@pytest.mark.skip_xp_backends('jax.numpy', reason=vf_skip_reason)
```
[the remark this answers is not in the document]
Comment (mdhaber) [reply scipy/ndimage/tests/test_filters.py:2773]: Ah, I didn't know `exceptions` worked for `np_only`; I've only used it for `cpu_only`.

File: scipy/ndimage/tests/test_filters.py [hunk 3]
```diff
@@ -2766,52 +2766,61 @@ def test_gh_22333():
...
     @pytest.mark.parametrize("dtype",
                              [np.uint8, np.uint16, np.uint32, np.uint64,
                               np.int8, np.int16, np.int32, np.int64,
                               np.float32, np.float64, np.complex64, np.complex128])
     @pytest.mark.parametrize("batch_memory", [1, 16*3, np.inf])
     @pytest.mark.parametrize("use_footprint", [False, True])
-    def test_dtype_batch_memory(self, dtype, batch_memory, use_footprint):
-        rng = np.random.default_rng(435982456983456987356)
+    def test_dtype_batch_memory(self, dtype, batch_memory, use_footprint, xp):
+        rng = xp.random.default_rng(435982456983456987356)
```
[the remark this answers is not in the document]
Comment (mdhaber) [reply scipy/ndimage/tests/test_filters.py:2816]: OK, yeah, if we are going to get this working (by converting back and forth) for other backends, I will change that.

File: scipy/ndimage/_filters.py (hunk 1 above)
Comment (mdhaber) [reply scipy/ndimage/_filters.py:77]: I can do that, but since `.lib.stride_tricks.sliding_window_view` is not close to array API compatible, I didn't really try to make the rest array API compatible (even where possible).

Comment (ev-br) [reply scipy/ndimage/_filters.py:77]: Yeah, let's at least avoid definitely non-existent things though.

I'd actually suggest doing something like 

```
if x.device == "cuda":
   stride_tricks = cupy.lib.stride_strics
else:
   stride_tricks = numpy.lib.stride_strics
....
# use stride_tricks as usual
```

and turn on testing with `array_api_strict`.

Comment (mdhaber) [reply scipy/ndimage/_filters.py:77]: I think you mean make the function array API compatible until that last step. Let's have a quick real-time chat if you have a moment? I want to make sure I understand the rest.

**Comment by mdhaber** [comment]
> So all array conversion logic is centralized. Now that I typed this all, I totally see why this could be a separate PR.

Yeah, I don't plan to touch that here. As the title suggests, I didn't intend to make the function fully array API compatible; just CuPy compatible. I'm already going quite a bit beyond that.
 
Actually, I had forgotten that the generic convert to numpy, do the work, convert back doesn't even work here. We have to pass the user `function` an array of the same dtype as the input, and generic machinery wouldn't do that. So I translated pretty much everything to the array API except `pad` and `sliding_window_view`. 

I added a bunch of conditional `pytest.skip`s inside the tests just to get things running. We can refactor the tests as you see fit. The main things are:
- Array types that aren't mutable obviously can't support `output`.
- Array types that aren't mutable don't support limited `batch_memory` yet. When `batch_memory` is limited, we loop over small chunks of the sliding window view and assign the results to an output array (even if `output` is not specified). If we can't mutate that output array, we either have to store all the separate pieces in a list and then `stack` them, or we would have to use `xpx.at.set` (and we'd end up copying the output array each loop iteration). Either you or Ilhan (or both) have not liked the "store chunks in list then stack" approach before... what should we do here?
- Using `footprint` currently requires an indexing operation like `[..., mask]` that Dask and array_api_strict don't support because the boolean index isn't the whole index. My initial attempt to get around that by broadcasting the mask and reshaping didn't work, but we can take another look at that later.

**Comment by mdhaber** [comment]
Note to improve in follow up, perhaps: if `size` is a scalar, it is turned into a tuple with the same number of dimensions as the input image. This is exactly as documented for all filters, but it is not ideal if `axes` are provided - it might be better if `size` became a tuple with the same number of elements as `axes`.

**Review by ev-br — COMMENTED** [review]
Looks fine overall. A suggested improvement looks good, too; can go either here or in a follow-up, whichever is more convenient.

CI failures with pytorch dtype mismatches are probably not too hard to fix.

File: scipy/ndimage/tests/test_filters.py [hunk 4]
```diff
@@ -2766,52 +2767,72 @@ def test_gh_22333():
...
     @pytest.mark.parametrize("use_footprint", [False, True])
-    def test_dtype_batch_memory(self, dtype, batch_memory, use_footprint):
+    def test_dtype_batch_memory(self, dtype, batch_memory, use_footprint, xp):
         rng = np.random.default_rng(435982456983456987356)
         w = 3
 
+        if is_jax(xp) and not (batch_memory == 1):
+            pytest.skip("Requires mutable array.")
+        if is_torch(xp) and dtype in {'uint8', 'uint16', 'uint32', 'uint64'}:
+            pytest.skip("Needs uint support.")
```
[the remark this answers is not in the document]
Comment (mdhaber) [reply scipy/ndimage/tests/test_filters.py:2825]: I think it was. I just didn't get fine-grained with the skip.

File: scipy/ndimage/_filters.py [hunk 5]
```diff
@@ -117,7 +120,7 @@ def footprinted_function(input, *args, axis=-1, **kwargs):
             raise ValueError(message)
 
     # mode must be one of the allowed strings, and we should convert it to the
-    # value required by `np.pad` here.
+    # value required by `xp.pad` here.
```
[the remark this answers is not in the document]
Comment (mdhaber) [reply scipy/ndimage/_filters.py:123]: ? You can see it's tested against other backends on the CPU. It just needs to resort to NumPy for the padding and stride tricks. No-copy conversion to the original backend is done before calling the user-defined callable.

Originally, in the first commit, the function only worked for CuPy and NumPy, but I generalized it, because you seemed to be asking for this:
https://github.com/scipy/scipy/pull/22715#discussion_r2003894917

[the remark this answers is not in the document]
Comment (ev-br) [reply scipy/ndimage/_filters.py:123]: The code comment says `xp.pad` which is, as you see, confusing. Easiest to just write np.pad, and the whole issue does not even arise.

File: scipy/ndimage/tests/test_filters.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (mdhaber) [reply scipy/ndimage/tests/test_filters.py:2827]: K, removed `uint8` fromt this list.

File: scipy/ndimage/_filters.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (mdhaber) [reply scipy/ndimage/_filters.py:123]: OK. Actually I removed a lot of the comment; it was out of date. The thing I was referring to happens in input validation now.

**Comment by mdhaber** [comment]
Fixed pytorch dtype mismatch; it was just mismatch between NumPy and torch default types in the test.

File: scipy/ndimage/tests/test_filters.py [hunk 6]
```diff
@@ -2909,44 +2936,46 @@ def test_input_validation(self):
...
             ndimage.vectorized_filter(input, function, size=size, batch_memory=0)
         with pytest.raises(ValueError, match=message):
             ndimage.vectorized_filter(input, function, size=size, batch_memory=(1, 2))
-        with pytest.raises(ValueError, match=message):
-            ndimage.vectorized_filter(input, function, size=size,
-                                      batch_memory="shrubbery")
+        with pytest.raises((ValueError, TypeError), match=message):
+            ndimage.vectorized_filter(input, function, size=size, batch_memory="a duck")
 
+    @pytest.mark.skip_xp_backends('torch', reason="Need to debug.")
```
Comment (ev-br) [inline scipy/ndimage/tests/test_filters.py:2953]: What to debug it here or kick the can down the road?

**Review by ev-br — APPROVED** [review]
LGTM.

While I'm still not a fan of `xp.pad` in the comments, but if there are strong feelings, fine, this isn't a deal-breaker.

File: scipy/ndimage/tests/test_filters.py (hunk 6 above)
Comment (mdhaber) [reply scipy/ndimage/tests/test_filters.py:2953]: I'll take a quick look this afternoon and at least provide a more informative reason. 

I didn't remember that there were other appearances of "xp.pad"? If others remain, I'll clarify that it's NumPy or CuPy.

Comment (mdhaber) [reply scipy/ndimage/tests/test_filters.py]: Fixed. Just a PyTorch `size` gotcha.

**Comment by ev-br** [comment]
The remaining CI failure looks unrelated. Thanks Matt.
